LKRG Trips kCFI During get_kallsyms_address (on-load)

Finally managed to get a 6.1.8 built with kCFI and LTO on LLVM16, LKRG still as a module. Unfortunately, the modules attempt to look-up addresses (it was built in-tree - in the hopes of avoiding this exact type of nonsense) at load-time produces this effect (hey, CFI works):
```sh
# modprobe lkrg 
[ 1285.469868][  T440] CFI failure at get_kallsyms_address+0xe6/0x150 [lkrg] (target: kallsyms_lookup_name+0x4/0x210; expected type: 0xaa7e236f)
[ 1285.471017][  T440] invalid opcode: 0000 [#1] PREEMPT SMP
[ 1285.471475][  T440] CPU: 3 PID: 440 Comm: modprobe Tainted: G                TN 6.1.8 #1
[ 1285.472168][  T440] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS Arch Linux 1.16.1-1-1 04/01/2014
[ 1285.472984][  T440] RIP: 0010:get_kallsyms_address+0xe6/0x150 [lkrg]
[ 1285.473570][  T440] Code: 28 48 89 05 f4 7c 02 00 48 89 e7 e8 44 e4 81 ec 4c 8b 1d e5 7c 02 00 48 c7 c7 41 06 9a c0 41 ba 91 dc 81 55 45 03 53 fc 74 02 <0f> 0b 41 ff d3 66 90 48 89 05 74 7d 02 00 31 db eb 10 48 c7 c3 ff
[ 1285.475287][  T440] RSP: 0018:ffffb18881bbf998 EFLAGS: 00010202
[ 1285.475863][  T440] RAX: 0000000000000000 RBX: ffffffffc0911005 RCX: 0000000000000000
[ 1285.476512][  T440] RDX: 0000000000384a83 RSI: 0000000000000000 RDI: ffffffffc09a0641
[ 1285.477155][  T440] RBP: ffffb18881bbfd70 R08: ffff97d9847ee0c0 R09: ffff97d980042800
[ 1285.477834][  T440] R10: 0000000055a0ebf7 R11: ffffffffad176a94 R12: ffffffffaea19728
[ 1285.478500][  T440] R13: 0000000000000000 R14: ffff97d9848e02a0 R15: 0000000000000000
[ 1285.479161][  T440] FS:  000066380b197740(0000) GS:ffff97dab7cc0000(0000) knlGS:0000000000000000
[ 1285.479924][  T440] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 1285.480463][  T440] CR2: 000066380aaf4400 CR3: 000000010a1da001 CR4: 0000000000760ee0
[ 1285.481115][  T440] PKRU: 55555554
[ 1285.481396][  T440] Call Trace:
[ 1285.481654][  T440]  <TASK>
[ 1285.481894][  T440]  ? kallsyms_lookup_name+0x4/0x210
[ 1285.482309][  T440]  ? __cfi_p_tmp_kprobe_handler+0x10/0x10 [lkrg]
[ 1285.482873][  T440]  ? 0xffffffffc04c7000
[ 1285.483227][  T440]  ? __cfi_init_module+0x5/0x5 [lkrg]
[ 1285.483674][  T440]  init_module+0x17/0xffb [lkrg]
[ 1285.484105][  T440]  ? __cfi_init_module+0x5/0x5 [lkrg]
[ 1285.484554][  T440]  do_one_initcall+0x12d/0x2b0
[ 1285.484959][  T440]  ? rcu_nocb_gp_kthread+0x470/0x9f0
[ 1285.485384][  T440]  do_init_module+0x51/0x280
[ 1285.485776][  T440]  __se_sys_finit_module+0xae/0x100
[ 1285.486204][  T440]  do_syscall_64+0x76/0xb0
[ 1285.486575][  T440]  ? do_syscall_64+0x87/0xb0
[ 1285.486998][  T440]  ? syscall_exit_to_user_mode+0x32/0x160
[ 1285.487494][  T440]  ? do_syscall_64+0x87/0xb0
[ 1285.487914][  T440]  ? do_syscall_64+0x87/0xb0
[ 1285.488325][  T440]  ? exc_page_fault+0x66/0xc0
[ 1285.488750][  T440]  entry_SYSCALL_64_after_hwframe+0x63/0xcd
[ 1285.489276][  T440] RIP: 0033:0x66380ab19abd
[ 1285.489673][  T440] Code: 5d c3 66 2e 0f 1f 84 00 00 00 00 00 90 f3 0f 1e fa 48 89 f8 48 89 f7 48 89 d6 48 89 ca 4d 89 c2 4d 89 c8 4c 8b 4c 24 08 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d 9b 72 0d 00 f7 d8 64 89 01 48
[ 1285.491451][  T440] RSP: 002b:000077aba59dc228 EFLAGS: 00000246 ORIG_RAX: 0000000000000139
[ 1285.492218][  T440] RAX: ffffffffffffffda RBX: 00000305cddba1d0 RCX: 000066380ab19abd
[ 1285.492904][  T440] RDX: 0000000000000000 RSI: 00000305b664acb2 RDI: 0000000000000003
[ 1285.493591][  T440] RBP: 00000305b664acb2 R08: 0000000000000000 R09: 0000000000000000
[ 1285.494315][  T440] R10: 0000000000000003 R11: 0000000000000246 R12: 0000000000040000
[ 1285.495336][  T440] R13: 00000305cddba160 R14: 0000000000000000 R15: 00000305cddba4c0
[ 1285.496361][  T440]  </TASK>
[ 1285.496878][  T440] Modules linked in: lkrg(+) intel_rapl_msr intel_rapl_common intel_uncore_frequency_common isst_if_common nfit kvm_intel kvm irqbypass crct10dif_pclmul polyval_clmulni polyval_generic gf128mul ghash_clmulni_intel qxl rapl drm_ttm_helper ttm joydev cfg80211 mousedev drm_kms_helper tiny_power_button sysimgblt evdev syscopyarea input_leds rfkill sysfillrect pcspkr fb_sys_fops uio_pdrv_genirq led_class i2c_piix4 button uio psmouse mac_hid sch_fq_codel fuse drm dmi_sysfs qemu_fw_cfg ip_tables x_tables ext4 crc32c_generic mbcache crc16 jbd2 virtio_net net_failover virtio_blk failover virtio_balloon ata_generic serio_raw crc32_pclmul crc32c_intel atkbd sha512_ssse3 uhci_hcd pata_acpi vivaldi_fmap aesni_intel ehci_pci libps2 crypto_simd ehci_hcd cryptd i8042 usbcore virtio_pci virtio_pci_legacy_dev floppy ata_piix usb_common serio virtio_pci_modern_dev
[ 1285.506399][  T440] ---[ end trace 0000000000000000 ]---
[ 1285.507383][  T440] RIP: 0010:get_kallsyms_address+0xe6/0x150 [lkrg]
[ 1285.508755][  T440] Code: 28 48 89 05 f4 7c 02 00 48 89 e7 e8 44 e4 81 ec 4c 8b 1d e5 7c 02 00 48 c7 c7 41 06 9a c0 41 ba 91 dc 81 55 45 03 53 fc 74 02 <0f> 0b 41 ff d3 66 90 48 89 05 74 7d 02 00 31 db eb 10 48 c7 c3 ff
[ 1285.511014][  T440] RSP: 0018:ffffb18881bbf998 EFLAGS: 00010202
[ 1285.511740][  T440] RAX: 0000000000000000 RBX: ffffffffc0911005 RCX: 0000000000000000
[ 1285.512602][  T440] RDX: 0000000000384a83 RSI: 0000000000000000 RDI: ffffffffc09a0641
[ 1285.513466][  T440] RBP: ffffb18881bbfd70 R08: ffff97d9847ee0c0 R09: ffff97d980042800
[ 1285.514408][  T440] R10: 0000000055a0ebf7 R11: ffffffffad176a94 R12: ffffffffaea19728
[ 1285.515326][  T440] R13: 0000000000000000 R14: ffff97d9848e02a0 R15: 0000000000000000
[ 1285.516246][  T440] FS:  000066380b197740(0000) GS:ffff97dab7cc0000(0000) knlGS:0000000000000000
[ 1285.517236][  T440] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 1285.518069][  T440] CR2: 000066380aaf4400 CR3: 000000010a1da001 CR4: 0000000000760ee0
[ 1285.518984][  T440] PKRU: 55555554
Segmentation fault
```
on the bright side, ought to mess up some (less friendly) rootkits this way :wink:

@Adam-pi3 - suggest trying to build with LLVM16 - it warns quite a bit about:
```
security/lkrg/modules/exploit_detection/syscalls/compat/p_compat_sys_keyctl/../../../../../modules/exploit_detection/syscalls/p_security_ptrace_access/p_security_ptrace_access.h:21:9: warning: 'P_LKRG_EXPLOIT_DETECTION_SECURITY_PTRACE_ACCESS_H' is used as a header guard here, followed by #define of a different macro [-Wheader-guard]
```
and _tons_ of:
```
security/lkrg/modules/exploit_detection/syscalls/compat/p_compat_sys_keyctl/../../../../../modules/exploit_detection/p_exploit_detection.h:498:16: warning: no case matching constant switch condition '0'
```

> Thanks. Well, it just produces a warning, right? Maybe we should add `noinline` and it'd work in practice?

It is just a warning, yes, but it seems like an orphaned section is being produced:

```
ld.lld: error: vmlinux.o:(.p_lkrg_read_only) is being placed in '.p_lkrg_read_only'
make[2]: *** [scripts/Makefile.vmlinux:91: vmlinux.unstripped] Error 1
make[1]: *** [/usr/src/linux-6.15.11-gentoo/Makefile:1236: vmlinux] Error 2
make: *** [Makefile:248: __sub-make] Error 2
```

Also, without `noinline`, `objcopy` complains:
```
vmlinux.o: warning: objtool: p_uninit_page_attr+0x96: stack state mismatch: cfa1=4+16 cfa2=4+48
vmlinux.o: warning: objtool: p_lkrg_register+0x462: return with modified stack frame
vmlinux.o: warning: objtool: p_lkrg_deregister+0x44: return with modified stack frame
```

--- a/CHANGES
+++ b/CHANGES
@@ -6,6 +6,7 @@ The following major changes have been made since 0.9.9:
     for cred pointer overwrites in certain other places where we did not before
  *) Do not track those credentials that we currently do not validate anyway
  *) Support (or rather be compatible with the kernel's use of) Intel CET IBT
+    (CONFIG_X86_KERNEL_IBT) and/or KCFI (CONFIG_CFI_CLANG) for now on x86_64
  *) Switch many hooks from kretprobes to simple kprobes for greater reliability
     and improved performance
  *) Overhaul locking of per-task shadow data, using finer-grain locks

--- a/src/p_lkrg_main.h
+++ b/src/p_lkrg_main.h
@@ -324,7 +324,7 @@ extern p_ro_page p_ro;
 #define P_CTRL(p_field) p_ro.p_lkrg_global_ctrl.ctrl.p_field
 #define P_CTRL_ADDR &p_ro.p_lkrg_global_ctrl
 
-#ifdef CONFIG_X86_KERNEL_IBT
+#if defined(CONFIG_X86_KERNEL_IBT) || defined(CONFIG_CFI_CLANG)
 
 /*
  * When Intel CET IBT is in use, we can't call non-exported kernel functions
@@ -338,10 +338,10 @@ extern p_ro_page p_ro;
  * We use macros to generate wrapper functions for making CET IBT compatible
  * indirect calls via PUSH/RET, one wrapper function per function pointer.
  *
- * We rely on these functions having no prologue/epilogue, but having a return
- * instruction or a branch to a return thunk.  __attribute__ ((naked)) would
- * more reliably avoid prologue/epilogue, but then it'd be our job to invoke
- * the right return thunk for a given build.
+ * With gcc, we rely on these functions having no prologue/epilogue, but having
+ * a return instruction or a branch to a return thunk.  __attribute__ ((naked))
+ * would more reliably avoid prologue/epilogue, but then it'd be our job to
+ * invoke the right return thunk for a given build.
  *
  * It is important to prevent inlining and other interprocedural optimizations
  * at least because the indirect function call isn't exposed to the compiler,
@@ -354,10 +354,33 @@ extern p_ro_page p_ro;
  * support.  In testing, noinline along with noclone wasn't good enough.  We
  * could improve upon this by having these functions in their own translation
  * unit, ideally coming from an assembly rather than C source file.
+ *
+ * With clang not supporting __attribute__ ((noipa)), we use a combination of
+ * attributes including __attribute__ ((naked)), which means that for CET IBT
+ * (unlike solely clang software CFI) we use an explicit return instruction,
+ * which unfortunately produces this sort of objtool warnings:
+ * vmlinux.o: warning: objtool: call_p_kallsyms_lookup_name+0xa: 'naked' return found in MITIGATION_RETHUNK build
  */
+
+#ifdef CONFIG_X86_64
+ #ifdef __clang__
+  #define GENERATE_CALL_FUNC_ATTR __attribute__ ((noduplicate)) __attribute__ ((naked)) __attribute__ ((optnone))
+  #ifdef CONFIG_X86_KERNEL_IBT
+   #define GENERATE_CALL_FUNC_ASM "pushq %0\n\tret"
+  #else
+   #define GENERATE_CALL_FUNC_ASM "jmpq *%0"
+  #endif
+ #else /* gcc */
+  #define GENERATE_CALL_FUNC_ATTR __attribute__ ((noipa))
+  #define GENERATE_CALL_FUNC_ASM "pushq %0"
+ #endif
+#else
+#error "CONFIG_X86_KERNEL_IBT and/or CONFIG_CFI_CLANG not yet supported on this architecture"
+#endif
+
 #define GENERATE_CALL_FUNC(type, name, ...) \
-   notrace __attribute__ ((noipa)) type call_##name(__VA_ARGS__) { \
-      __asm__ __volatile__ ("pushq %0" :: "m" (P_SYM(name))); \
+   notrace noinline GENERATE_CALL_FUNC_ATTR type call_##name(__VA_ARGS__) { \
+      __asm__ __volatile__ (GENERATE_CALL_FUNC_ASM :: "m" (P_SYM(name))); \
    } \
    STACK_FRAME_NON_STANDARD(call_##name);
 

